// whatever-storage/lib/whateverStorage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _window = _interopRequireDefault(require("global/window"));

var _StorageFallback = _interopRequireDefault(require("./StorageFallback"));

var _cookieJar = _interopRequireDefault(require("./cookieJar"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function whateverStorage(storage) {
  if (storage === 'cookies') {
    return _cookieJar.default;
  }

  var storageMethod;

  try {
    _window.default[storage].setItem('test', 'test');

    _window.default[storage].removeItem('test');

    storageMethod = _window.default[storage];
  } catch (err) {
    storageMethod = new _StorageFallback.default();
  }

  return {
    getItem: function getItem(key) {
      var json = storageMethod.getItem(key) || null;
      return JSON.parse(json);
    },
    setItem: function setItem(key) {
      var value = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      storageMethod.setItem(key, JSON.stringify(value));
    },
    removeItem: function removeItem(key) {
      storageMethod.removeItem(key);
    },
    clear: function clear() {
      storageMethod.clear();
    },
    key: function key() {
      var index = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
      return storageMethod.key(index);
    }
  };
}

var _default = {
  local: whateverStorage('localStorage'),
  session: whateverStorage('sessionStorage'),
  cookie: whateverStorage('cookies')
};
exports.default = _default;
